libraries/AP_HAL_ChibiOS/RCOutput.cpp

```diff
@@ -253,14 +253,23 @@ void RCOutput::dshot_collect_dma_locks(uint32_t time_out_us)
             // if we have time left wait for the event
             eventmask_t mask = 0;
             const uint32_t pulse_elapsed_us = now - group.last_dmar_send_us;
+            uint32_t wait_us = 0;
             if (now < time_out_us) {
-                mask = chEvtWaitOneTimeout(group.dshot_event_mask,
-                    chTimeUS2I(MAX(time_out_us - now, group.dshot_pulse_send_time_us - pulse_elapsed_us)));
+                wait_us = MAX(time_out_us - now, group.dshot_pulse_send_time_us - pulse_elapsed_us);
             } else if (pulse_elapsed_us < group.dshot_pulse_send_time_us) {
                 // better to let the burst write in progress complete rather than cancelling mid way through
-                mask = chEvtWaitOneTimeout(group.dshot_event_mask,
-                    chTimeUS2I(group.dshot_pulse_send_time_us - pulse_elapsed_us));
+                wait_us = group.dshot_pulse_send_time_us - pulse_elapsed_us;
             }
+
+            // waiting for a very short period of time can cause a
+            // timer wrap with ChibiOS timers. Use CH_CFG_ST_TIMEDELTA
+            // as minimum. Don't allow for a very long delay (over 1200us)
+            // to prevent bugs in handling timer wrap
+            const uint32_t max_delay_us = 1200;
```

Comment:
this should be _dshot_period_us

Reply:
i'm not sure it should be. _dshot_period_us is the time between when we start DShot pulses. This is the max time we wait for a DMA to complete. It isn't clear why they should be related. 
The non-contended case for dshot is dshot_bit_length pulses, which is max 127 usec for DShot 150. If a DShot DMA takes much longer than 127usec then it really must be due to DMA contention. Using _dshot_period_us would give 1250us for 800Hz, but I'm not convinced we really want to wait 1250us.
We don't want to cancel a DMA operation unless we think it really won't complete. Cancelling an operation can corrupt the pulses.
How often do we actually hit this timeout? Maybe we should just have a fixed timeout so we don't block forever on loss of a DMA completion?

Reply:
The DMA lock is taken at the beginning of the pulse and is released on the return path (so for bdshot after the return pulse has come back). The timeout is currently set to the end of the period, this is because we want to stay in sync with push() and because we may issue multiple pulses for one push() so for instance if we run the loop rate at 800Hz and SERVO_DSHOT_RATE at 3 the dshot interval will be for 2.4Khz - so a fixed timeout of 1200 will be much too long - it will overlap with multiple pulses.

Reply:
Please note that I tried multiple values for this on Saturday in an attempt to make your patch work - all values were broken in some way, so it's critical that I test this on my hardware before it goes in.

Reply:
it would be worth trying a fixed large timeout. We really should be hitting the timeout only extremely rarely, under very high levels of DMA contention, and cancelling them restarting doesn't really help us get uniform pulse period. So a fixed (and quite large) timeout is perhaps the simplest solution

Reply:
To be happy I will want to put on my Saleae and check with different configurations - that's what I was doing on Saturday

Reply:
note that we really do need this PR in soon. It is a fall out of the sky bug, so really critical. I'd rather have uneven dshot than watchdogs and complete vehicle loss.